// ton-contract-executor/dist/utils/parseActionList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseActionsList = void 0;
const ton_1 = require("ton");
const messageUtils_1 = require("./messageUtils");
function parseActionsListImpl(actions) {
    let list = [];
    let ref;
    let outAction;
    let slice;
    if (actions instanceof ton_1.Cell) {
        slice = ton_1.Slice.fromCell(actions);
    }
    else {
        slice = actions;
    }
    try {
        ref = slice.readRef();
    }
    catch (e) {
        return list;
    }
    let magic = slice.readUint(32).toNumber();
    if (magic === 0x0ec3c86d) {
        outAction = {
            type: 'send_msg',
            mode: slice.readUint(8).toNumber(),
            message: (0, messageUtils_1.readMessage)(slice.readRef())
        };
    }
    else if (magic === 0x36e6b809) {
        outAction = {
            type: 'reserve_currency',
            mode: slice.readUint(8).toNumber(),
            currency: (0, messageUtils_1.readCurrencyCollection)(slice)
        };
    }
    else if (magic === 0xad4de08e) {
        outAction = {
            type: 'set_code',
            newCode: slice.readRef().toCell()
        };
    }
    else {
        outAction = { type: 'unknown' };
    }
    list.push(outAction);
    list.push(...parseActionsListImpl(ref));
    return list;
}
function parseActionsList(actions) {
    return parseActionsListImpl(actions).reverse();
}
exports.parseActionsList = parseActionsList;
